### PR TITLE
Update platforms.py  "superuser privileges" check

### PR DESCRIPTION
**Pull request by namloc2001**
## Description
To fix issue #6579. Making it so that in `platforms.py` the check for "superuser privileges" is only looking for uid/euid=0, not including the root group (gid/egid=0) which does not have the superuser privileges associated with uid/euid=0.

### CODE REVIEW

**Comment by namloc2001** [comment]
@auvipy I don't suppose you know why the tests would be failing, given this is such a minor change?

```
    c:\python37-x64\include\pyport.h(6) : fatal error C1083: Cannot open include file: 'inttypes.h': No such file or directory
    
        =============================DEBUG ASSISTANCE=============================
        If you are seeing a compilation error please try the following steps to
        successfully install cryptography:
        1) Upgrade to the latest pip and try again. This will fix errors for most
           users. See: https://pip.pypa.io/en/stable/installing/#upgrading-pip
        2) Read https://cryptography.io/en/latest/installation.html for specific
           instructions for your platform.
        3) Check our frequently asked questions for more information:
           https://cryptography.io/en/latest/faq.html
        =============================DEBUG ASSISTANCE=============================
    
    error: command 'c:\\Program Files (x86)\\Microsoft Visual Studio 10.0\\VC\\Bin\\amd64\\cl.exe' failed with exit status 2
    
    ----------------------------------------
Command "C:\projects\celery\.tox\3.7-unit\Scripts\python.EXE -u -c "import setuptools, tokenize;__file__='C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\pip-install-tcc2n6j9\\cryptography\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\appveyor\AppData\Local\Temp\1\pip-record-umzf7tol\install-record.txt --single-version-externally-managed --compile --install-headers C:\projects\celery\.tox\3.7-unit\include\site\python3.7\cryptography" failed with error code 1 in C:\Users\appveyor\AppData\Local\Temp\1\pip-install-tcc2n6j9\cryptography\
You are using pip version 10.0.1, however version 20.3.3 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
=================================== log end ===================================
ERROR: could not install deps [-rC:\projects\celery/requirements/default.txt, -rC:\projects\celery/requirements/test.txt, -rC:\projects\celery/requirements/pkgutils.txt, -rC:\projects\celery/requirements/test-ci-default.txt, -rC:\projects\celery/requirements/docs.txt, -rC:\projects\celery/requirements/docs.txt]; v = InvocationError("'C:\\projects\\celery\\.tox\\3.7-unit\\Scripts\\python.EXE' -m pip install '-rC:\\projects\\celery/requirements/default.txt' '-rC:\\projects\\celery/requirements/test.txt' '-rC:\\projects\\celery/requirements/pkgutils.txt' '-rC:\\projects\\celery/requirements/test-ci-default.txt' '-rC:\\projects\\celery/requirements/docs.txt' '-rC:\\projects\\celery/requirements/docs.txt'", 1)
___________________________________ summary ___________________________________
ERROR:   3.7-unit: could not install deps [-rC:\projects\celery/requirements/default.txt, -rC:\projects\celery/requirements/test.txt, -rC:\projects\celery/requirements/pkgutils.txt, -rC:\projects\celery/requirements/test-ci-default.txt, -rC:\projects\celery/requirements/docs.txt, -rC:\projects\celery/requirements/docs.txt]; v = InvocationError("'C:\\projects\\celery\\.tox\\3.7-unit\\Scripts\\python.EXE' -m pip install '-rC:\\projects\\celery/requirements/default.txt' '-rC:\\projects\\celery/requirements/test.txt' '-rC:\\projects\\celery/requirements/pkgutils.txt' '-rC:\\projects\\celery/requirements/test-ci-default.txt' '-rC:\\projects\\celery/requirements/docs.txt' '-rC:\\projects\\celery/requirements/docs.txt'", 1)
```

**Comment by auvipy** [comment]
> fatal error C1083: Cannot open include file: 'inttypes.h': No such file or directory

fatal error C1083: Cannot open include file: 'inttypes.h': No such file or directory

our windows CI is not perfect & we have to adjust some settings in our github actions migration from travis. it now hangs while running the integration tests

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:785]: Per https://github.com/celery/celery/issues/6579#issuecomment-761858980, we may wish to check if the group's name is `wheel` or figure out which unix systems still rely on gid 0.
Celery supports a wide variety of deployment targets, even if not officially. People have production systems that we're unaware of.
We should make the effort to ensure no security issues occur from this bugfix.

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py:785]: @thedrow how about we look at the group name as well as performing a UID/EUID check. Then if the group name is in a list of potential superuser privilege groups, we give the "`C_FORCE_ROOT`" related message and actions? Here's some sample code I've knocked up to demonstrate GID=1000 (a user defined group), GID=0 (root - which we'll allow) and GID=27 (sudo on my ubuntu instance).

```
#!/usr/bin/env python
# Python program to use grp.getgrgid() method to get group name from id. 
    
# import modules
import grp
import sys

# User and Effective User id
uid = 1000
euid = 1000

# Group id 
gid = 1000
egid = 1000

# Get the group database entry for the specified group id using grp.getgrgid() method
entry = grp.getgrgid(gid)

# Print the retrieved entry 
print("Group database entry for group id " + str(gid) + ": " + str(entry))

# Get and print group name based on gid
grp_entry = grp.getgrgid(gid)
grp_name = grp_entry[0]
print("Group Name for gid=" + str(gid) + ": " + grp_name)

if not uid or not euid:
    sys.stdout.write('User or Effective User ID is zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
    sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')
else:
    if str(grp_name) == 'sudo' or str(grp_name) == 'wheel':
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + ', however group is: ' + grp_name + '\n')
        sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')		
    else:
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
        sys.stdout.write('You are good to go!\n')

# =========================================================================================
# =========================================================================================
# =========================================================================================

# Group id 
gid = 0 # group is root
egid = 0
  
# Get the group database entry for the specified group id using grp.getgrgid() method 
entry = grp.getgrgid(gid) 

# Print the retrieved entry 
print("\nGroup database entry for group id " + str(gid) + ": " + str(entry))

# Get and print group name based on gid
grp_entry = grp.getgrgid(gid)
grp_name = grp_entry[0]
print("Group Name for gid=" + str(gid) + ": " + grp_name)

if not uid or not euid:
    sys.stdout.write('User or Effective User ID is zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
    sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')
else:
    if str(grp_name) == 'sudo' or str(grp_name) == 'wheel':
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + ', however group is: ' + grp_name + '\n')
        sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')	
    else:
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
        sys.stdout.write('You are good to go!\n')

# =========================================================================================
# =========================================================================================
# =========================================================================================

# Group id 
gid = 27 # group is sudo
egid = 27

# Get the group database entry for the specified group id using grp.getgrgid() method
entry = grp.getgrgid(gid) 

# Print the retrieved entry 
print("\nGroup database entry for group id " + str(gid) + ": " + str(entry))

# Get and print group name based on gid
grp_entry = grp.getgrgid(gid)
grp_name = grp_entry[0]
print("Group Name for gid=" + str(gid) + ": " + grp_name)

if not uid or not euid:
    sys.stdout.write('User or Effective User ID is zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
    sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')
else:
    if str(grp_name) == 'sudo' or str(grp_name) == 'wheel':
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + ', however group is: ' + grp_name + '\n')
        sys.stdout.write('Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!\n')	
    else:
        sys.stdout.write('No user IDs are zero: uid=' + str(uid) + ' euid=' + str(euid) + ' gid=' + str(gid) + ' egid=' + str(egid) + '\n')
        sys.stdout.write('You are good to go!\n')

# =========================================================================================
# =========================================================================================
# =========================================================================================
```

The output of this script on my Ubuntu machine is:
```
Group database entry for group id 1000: grp.struct_group(gr_name='matt', gr_passwd='x', gr_gid=1000, gr_mem=[])
Group Name for gid=1000: matt
No user IDs are zero: uid=1000 euid=1000 gid=1000 egid=1000
You are good to go!

Group database entry for group id 0: grp.struct_group(gr_name='root', gr_passwd='x', gr_gid=0, gr_mem=[])
Group Name for gid=0: root
No user IDs are zero: uid=1000 euid=1000 gid=0 egid=0
You are good to go!

Group database entry for group id 27: grp.struct_group(gr_name='sudo', gr_passwd='x', gr_gid=27, gr_mem=['matt'])
Group Name for gid=27: sudo
No user IDs are zero: uid=1000 euid=1000 gid=27 egid=27, however group is: sudo
Running a worker with superuser privileges when the worker accepts messages serialized with pickle is a very bad idea!
```

And I think this then gives the ability to define a list of groups by name (rather than ID which can be different based on distribution), that we are or aren't happy with as default and so require `C_FORCE_ROOT` override.

(The code might need some refactoring to also check `egid` against the list of named groups as well, but in principle it'll follow the same check as against `gid`).

Thoughts?

Comment (thedrow) [reply celery/platforms.py:785]: That should do the trick.
Note that these functions may raise a `KeyError` and in that case we should failover to mandating `C_FORCE_ROOT` with an appropriate log message.

I'm being particularly meticulous here to avoid potential security issues.

Comment (namloc2001) [reply celery/platforms.py:785]: Yep, can completely understand and appreciate that.

Comment (namloc2001) [reply celery/platforms.py:793]: @thedrow I've updated the code. Can you please explain where you feel the `KeyError` may come in and specify the code you want to handle it so I can include it?

Comment (auvipy) [reply celery/platforms.py:797]: He probably metioned def performTest(uid,euid,gid,egid):
    # Get the group database entry for the specified group id using grp.getgrgid() method
    try:
        entry = grp.getgrgid(gid)
    except KeyError:
        print('gid or egid not found')
        return  # 'exit' function and return to caller

and suggested to fallback to or failover to mandating C_FORCE_ROOT with an appropriate log message.

Comment (namloc2001) [reply celery/platforms.py:816]: @thedrow and @auvipy I've added in the `KeyError` handler please can you review.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:792]: Use the logger. Calling `warnings.warn()` will redirect the message to the logger.
We can define a new type of warning called `SecurityWarning`.
Prints are not our usual API to output warning messages to the screen.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:797]: Use the logger here to emit an error.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:803]: Let's change this to `SecurityWarning`.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:816]: Instead of `not uid or not euid` we should just check equality to 0.
We can extract the check to the `is_root` variable so that the comment will be redundant and thus we will be able to remove it.

That is unless there's a chance either `uid` or `euid` can be None.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:832]:
```suggestion
    elif any(x in gid_list for x in group_list):
```

Python accepts generators without parenthesis as arguments.
Furthermore, the condition is redundant.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:812]:
```suggestion
    gid_list = (gid_grp_name, egid_grp_name)
```
This is constant. It should be a tuple as it is faster to create.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:813]:
```suggestion
    group_list = ('sudo', 'wheel')
```

This is constant. It should be a tuple as it is faster to create.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:787]:
```suggestion
    # We must handle the case where either the gid or the egid are not found.
```

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:794]: We're testing this twice. Both here and in line 817-818.
We should extract the test to a variable.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:818]: Extract to a variable, see above.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:821]: Use the error logger.

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:783]:
```suggestion
                   for attr in ('getuid', 'getgid', 'geteuid', 'getegid')):
```

If we're reviewing this part of the code, we might as well improve it.

**Review by thedrow — CHANGES_REQUESTED** [review]
The rest of my comments are mostly nitpicks.
After we address those, this is good to go.

File: celery/platforms.py
Comment (thedrow) [reply celery/platforms.py:832]: Also, the variable names are not as meaningful as they could be.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: @thedrow I've amended the code for this suggestion. Can you please verify the syntax I've used? Does `SecurityWarning` need to be defined elsewhere?

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: Have amended, please verify updated code is acceptable.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: Have amended, please verify updated code is acceptable.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: Have amended, please verify updated code is acceptable.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: Variables renamed as well.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: @thedrow done, please check

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: Done, please check

File: celery/platforms.py
Comment (thedrow) [inline celery/platforms.py:776]:
```suggestion
    pickle_or_serialize = 'pickle' in accept_content or 'application/x-python-serialize' in accept_content
```

This already evaluates to a boolean.

File: celery/platforms.py
Comment (thedrow) [reply celery/platforms.py:817]: @namloc2001 Please address this too.

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/platforms.py
Comment (thedrow) [reply celery/platforms.py]: This should be an error. Not a warning.
Use the logger object.
You can get it using [`get_logger()`](https://docs.celeryproject.org/en/stable/internals/reference/celery.utils.log.html#celery.utils.log.get_logger).

File: celery/platforms.py
Comment (thedrow) [reply celery/platforms.py]: This should be an error. Not a warning.
Use the logger object.
You can get it using [`get_logger()`](https://docs.celeryproject.org/en/stable/internals/reference/celery.utils.log.html#celery.utils.log.get_logger).

**Comment by thedrow** [comment]
Before merging this PR we also need to fully unit test the code.
Please ensure the tests are up to date and are passing.
We're still working on our new CI.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: @thedrow sorry, out of my depth here, I don't know how to do this.

File: celery/platforms.py
Comment (namloc2001) [reply celery/platforms.py]: @thedrow I've updated the code please review.

**Comment by thedrow** [comment]
Looks OK, what about the tests?

**Comment by namloc2001** [comment]
@thedrow what about the logger piece? I couldn't work out what you needed here?

File: celery/platforms.py
Comment (thedrow) [reply celery/platforms.py]: I'll handle it.

**Comment by thedrow** [comment]
@namloc2001 I can't push to your branch since you didn't [allow edits from a maintainer](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
I'm working on this and I have begun writing tests. Can you please check the box so we can collaborate?

**Comment by namloc2001** [comment]
> @namloc2001 I can't push to your branch since you didn't [allow edits from a maintainer](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
> I'm working on this and I have begun writing tests. Can you please check the box so we can collaborate?

Hi, allow edits by maintainers is ticked on this PR already. I've also invited you as a collaborator to the branch in case that is needed

**Comment by thedrow** [comment]
@auvipy Is codecov set up in CI?

**Comment by thedrow** [comment]
I'm marking this as a draft until we cover all code paths.

**Comment by auvipy** [comment]
> @auvipy Is codecov set up in CI?

it wasn't. I had internet issues for the last 2 days. so couldn't reply in time, sorry, but you set it up in the meantime. thanks

**Review by thedrow — APPROVED** [review]

**Pull request merged**
